Complete sprint 16

Evo-Eclipse opened the pull request:
Onboarding Flow

- Added a two-page onboarding experience using OnboardingViewController and OnboardingPageViewController
- Onboarding is shown only for first-time users; completion state is persisted in UserStore
- Updated SceneDelegate to choose initial root: onboarding for new users, main tab bar otherwise

Category Selection: MVVM Refactor

- Refactored TrackerCategorySelectionViewController to MVVM with TrackerCategorySelectionViewModel
- Moved category and selection state management from the controller into the view model
- Replaced inline text field creation with a dedicated modal: TrackerNewCategoryViewController
- Simplified empty state UI: placeholder stack view (image + label) when no categories are available
- Updated table configuration and bindings to reflect view model-driven updates

Tracker Creation Improvements

- Trimmed whitespace from tracker titles before validation and save to prevent empty/whitespace-only entries
- Updated category selection integration to instantiate TrackerCategorySelectionViewController with its ViewModel and store dependencies for consistent state

Architecture and Navigation

- Centralized category selection flow via view model construction to improve separation of concerns
- Removed direct category manipulation from view controllers in favor of view model actions and outputs
- Ensured smooth transition from onboarding to main interface upon completion

Evo-Eclipse committed:
Add page-based onboarding flow

- Implemented OnboardingViewController with horizontal paging
- Created OnboardingPageViewController for pages with image and text
- Updated SceneDelegate to present onboarding instead of SplashViewController
- Added onboarding images and UIPageControl; transition to TabBarViewController

Evo-Eclipse committed:
Add onboarding completion persistence

- Created UserStore singleton with UserDefaults-backed isOnboardingCompleted property
- Enhanced SceneDelegate to conditionally show TabBarViewController or OnboardingViewController based on completion state
- Updated OnboardingViewController to mark onboarding as completed on finish

Evo-Eclipse committed:
Refactor category selection with dedicated creation form

Extracted category creation into separate view controller and improved user experience flow

- Created TrackerNewCategoryViewController with text input validation and 38-character limit handling
- Refactored TrackerCategorySelectionViewController to show only existing categories with immediate selection
- Added placeholder stack view with icon and descriptive text for empty category list
- Replaced inline text field with "Add Category" button that presents modal creation flow
- Removed hardcoded default categories, now starts with empty list
- Enhanced selection flow to immediately return selected category and dismiss view controller

Evo-Eclipse committed:
Refactor TrackerCategorySelection to MVVM Architecture

Introduced MVVM pattern for category selection flow, separating concerns and improving testability

- Created TrackerCategorySelectionViewModel to manage category data, loading, selection, and updates via bindings and delegate protocol
- Updated TrackerCategorySelectionViewController to use ViewModel for data source, bindings (onCategoriesChanged, onSelectedCategoryChanged), and actions like addCategory and selectCategory
- Modified TrackerCreationFormViewController to instantiate category selection view controller with ViewModel, injecting TrackerCategoryStore from AppDelegate
- Enhanced TrackerCategoryStore with getAllCategories method and removed default category setup from TrackersViewController for better data handling
- Implemented TrackerCategoryStoreDelegate in ViewModel for reactive updates on Core Data changes

IgorChumikov commented on Tracker/Controllers/OnboardingPageViewController.swift:
```diff
@@ -0,0 +1,71 @@
+        return label
+    }()
+
+    // MARK: - Initializers
```
// MARK: - Init

IgorChumikov commented on Tracker/Controllers/OnboardingViewController.swift:
```diff
@@ -0,0 +1,163 @@
+
+        if let first = pages.first {
+            setViewControllers([first], direction: .forward, animated: true, completion: nil)
+        }
```
Твой viewDidLoad уже довольно аккуратный, но его можно сделать чище, вынеся инициализацию UIPageViewController в отдельный метод.
Например:
```

override func viewDidLoad() {
    super.viewDidLoad()
    configurePageViewController()
    setupViews()
    setupConstraints()
}

private func configurePageViewController() {
    dataSource = self
    delegate = self
    
    if let firstPage = pages.first {
        setViewControllers([firstPage], direction: .forward, animated: true)
    }
}

```

IgorChumikov commented on Tracker/Controllers/TrackerCreationFormViewController.swift:
```diff
@@ -510,10 +512,15 @@ extension TrackerCreationFormViewController: UITableViewDelegate {
     func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
         if indexPath.row == 0 {
-            let categoryVC = TrackerCategorySelectionViewController()
+            guard let appDelegate = UIApplication.shared.delegate as? AppDelegate else { return }
```
Класс AppDelegate можно сделать final class AppDelegate

IgorChumikov commented on Tracker/Application/SceneDelegate.swift:
```diff
@@ -19,7 +19,13 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+
         window?.makeKeyAndVisible()
     }
 
```
Класс SceneDelegate -->>> final class SceneDelegate

IgorChumikov commented on Tracker/Controllers/OnboardingViewController.swift:
```diff
@@ -0,0 +1,163 @@
+        // Persist completion so onboarding won't show next launches
+        UserStore.shared.isOnboardingCompleted = true
+
+        let tabBarViewController = TabBarViewController()
```
В классе TabBarViewController 2 раза import UIKit — это просто лишняя строка, её можно смело убрать.

IgorChumikov commented on Tracker/Resources/Assets.xcassets/Images/image_onboarding_first.imageset/Contents.json:
```diff
@@ -0,0 +1,22 @@
+{
```
Класс ViewController думаю можно удалить он просто пустой 
```

//
//  ViewController.swift
//  Tracker
//
//  Created by Pavel Komarov on 07.07.2025.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
    }


}


```

Evo-Eclipse committed:
Refactor project structure for style consistency

- Applied final keyword to system delegate classes
- Standardized MARK comments and method organization
- Cleaned up whitespace, import, and unused file
- Enhanced encapsulation with proper access control
